src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let stateCopy;
+
+    switch (action.type) {
+      case 'addProperties':
+        stateCopy = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        stateCopy = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+
+      case 'clear':
+        stateCopy = {};
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    currentState = stateCopy;
+    stateHistory.push({ ...currentState });
```

Comment:
The operation `{ ...currentState }` is redundant since `currentState` is already a copy of the state at this point. You can simply push `currentState` directly to `stateHistory`.